src/preamble.js

```diff
@@ -796,6 +796,75 @@ function createExportWrapper(name, fixedasm) {
 }
 #endif
 
+#if ABORT_ON_WASM_EXCEPTIONS
+// When DISABLE_EXCEPTION_CATCHING != 1 `abortWrapperDepth` counts the recursion 
+// level of the wrapper function so that we only handle exceptions at the top level
+// letting the exception mechanics work uninterrupted at the inner level.
+// Additionally, `abortWrapperDepth` is also manually incremented in callMain so that 
+// we know to ignore exceptions from there since they're handled by callMain directly.
+var abortWrapperDepth = 0;
+
+// Instrument all the exported functions to:
+// - abort if an unhandled exception occurs
+// - throw an exception if someone tries to call them after the program has aborted
+// See settings.ABORT_ON_WASM_EXCEPTIONS for more info.
+function instrumentWasmExportsWithAbort(exports) {
+  var instExports = {};
+  for (var name in exports) {
+    (function(name) {
+      var original = exports[name];
+      
+      // Wrap all functions, copy the other symbols.
+      if (typeof original === 'function') {
+        instExports[name] = function() {
+          // Don't allow this function to be called if we're aborted!
+          if (ABORT == 1) { 
+#if ASSERTIONS
+            throw "program has already aborted!";
+#else
+            throw "DEAD";	
+#endif
+          }
+          
+#if DISABLE_EXCEPTION_CATCHING != 1
+          abortWrapperDepth += 1;
+#endif
+          try {
+            return original.apply(null, arguments);
+          }
+          catch (e) {
+            if(
+              ABORT === 1 // rethrow exception if abort() was called in the original function call above
+              || abortWrapperDepth > 1 // rethrow exceptions not caught at the top level if exception catching is enabled; rethrow from exceptions from within callMain
+#if SUPPORT_LONGJMP
+              || e === 'longjmp' // rethrow longjmp if enabled
+#endif
+            ) {
+              throw e;
+            }
+            
+#if ASSERTIONS    
+            abort("unhandled exception: " + [e, e.stack]);
+#else
+            abort("UHE:" + [e, e.stack]);
```

Comment:
the stack etc. is excessive in a release build, I think, we really care about code size there.
```suggestion
            abort();
```

Reply:
I respectfully disagree :)

Without the stack trace it's impossible to find out where the crash happened in a release build because the only thing we'll get back from the field is that there was an abort (and the stack trace from abort()) with the callstack of the actual crash lost.

Getting the stack trace to the actual crash location is essential as there are many bugs we cannot reproduce locally or happen so rarely that it's impractical to do so. 

Since this feature is off-by-default I'm convinced anyone opting into it will gladly take the ~10 byte overhead in build size to get this information which quickly translates into huge real world value (customer satisfaction and qa/dev time).

Reply:
I see what you mean now. Makes sense!

In that case, how about removing the difference with assertions? It looks like the familiar pattern of "be super-concise when assertions are off" but isn't, which is what confused me. And it would make the code simpler. As you said, a few bytes in this special mode are not a big deal (but I'm fine with both modes having either the shorter or the longer error).